### PR TITLE
Show the expected stack trace from a rethrown exception.

### LINKED ISSUE #15780: The throw keyword shows the wrong thrown exception line
_From @NicolasDorier on August 22, 2017 14:56_

Using .NETCore2.0, Win10, it seems the `throw` keywor

Using the `throw` keyword loose all the initial stacktrace information. 
You can workaround with `ExceptionDispatchInfo.Capture(ex).Throw();`, but this seems like a bug to me.

```
using System;

namespace ConsoleApp3
{
	class Program
	{
		static void Main(string[] args)
		{
			try
			{
				DoStuff();
			}
			catch(Exception ex)
			{
				Console.WriteLine(ex.StackTrace);
			}
			
		}

		private static void DoStuff()
		{
			try
			{
				throw new Exception("Boom!");
			}
			catch(Exception ex)
			{
				throw;
			}
		}
	}
}
```
Actual output: (Exception line 28)

```
   at ConsoleApp3.Program.DoStuff() in c:\users\nicolasdorier\documents\visual studio 2017\Projects\ConsoleApp3\ConsoleApp3\Program.cs:line 28
   at ConsoleApp3.Program.Main(String[] args) in c:\users\nicolasdorier\documents\visual studio 2017\Projects\ConsoleApp3\ConsoleApp3\Program.cs:line 11
```

Expected output: (Exception line 24)

```
   at ConsoleApp3.Program.DoStuff() in c:\users\nicolasdorier\documents\visual studio 2017\Projects\ConsoleApp3\ConsoleApp3\Program.cs:line 24
   at ConsoleApp3.Program.Main(String[] args) in c:\users\nicolasdorier\documents\visual studio 2017\Projects\ConsoleApp3\ConsoleApp3\Program.cs:line 11
```

Workaround using `ExceptionDispatchInfo.Capture(ex).Throw();`:

```
using System;
using System.Runtime.ExceptionServices;

namespace ConsoleApp3
{
	class Program
	{
		static void Main(string[] args)
		{
			try
			{
				DoStuff();
			}
			catch(Exception ex)
			{
				Console.WriteLine(ex.StackTrace);
			}
			
		}

		private static void DoStuff()
		{
			try
			{
				throw new Exception("Boom!");
			}
			catch(Exception ex)
			{
				ExceptionDispatchInfo.Capture(ex).Throw();
				throw;
			}
		}
	}
}

```

_Copied from original issue: dotnet/corefx#23470_

### PR DESCRIPTION
**Pull request by ateoi**
Fix #15780

### CODE REVIEW

File: src/vm/excep.cpp [hunk 1]
```diff
@@ -2428,9 +2428,7 @@ void StackTraceInfo::SaveStackTrace(BOOL bAllowAllocMem, OBJECTHANDLE hThrowable
                         }
                     }
 
-                    if (bSkipLastElement && gc.stackTrace.Size() != 0)
-                        gc.stackTrace.AppendSkipLast(m_pStackTrace, m_pStackTrace + m_dFrameCount);
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/vm/excep.cpp]: Also: `StackTraceElement.PartiallyEqual/PartialAtomicUpdate`

**Comment by jkotas** [comment]
@ateoi Thanks for fixing this!

**Comment by danmosemsft** [comment]
Great to see a fix for this.

You might want to open a bug in the dotnet/docs repo to update [this text ](https://msdn.microsoft.com/en-us/library/system.exception.stacktrace(v=vs.110).aspx) for 2.1:
```
If the exception is thrown, and later rethrown, in the same method, the stack trace only contains the location where the exception was rethrown and does not include the location where the exception was originally thrown.
```

**Comment by danmosemsft** [comment]
Also, would you be willing to open a PR in the corefx repo to add a test? I am not certain the best place, I don't see tests specifically for what is in Exception.StackTrace. Some possible places are
src\System.Runtime\tests\System\Runtime\ExceptionServices\ExceptionDispatchInfoTests.netcoreapp.cs
src\System.Runtime.Extensions\tests\System\Environment.StackTrace.cs
src\System.Diagnostics.StackTrace\tests\StackTraceTests.cs
src\System.Runtime\tests\System\ExceptionTests.cs
A naive test could do something like
```c#
        static int expectedThrowLine;

        static void Main(string[] args)
        {
            try
            {
                DoStuff();
            }
            catch (Exception ex)
            {
                int reportedThrowLine = Convert.ToInt32(Regex.Match(ex.StackTrace, @".*DoStuff[^\n]*line (\d+)", RegexOptions.Multiline).Groups[1].Value);
                Assert.Equals(expectedThrowLine, reportedThrowLine);
            }

        }

        private static void DoStuff()
        {
            try
            {
                expectedThrowLine = 1 + Convert.ToInt32(Regex.Match(Environment.StackTrace, @".*DoStuff[^\n]*line (\d+)", RegexOptions.Multiline).Groups[1].Value);
                throw new Exception("Boom!");
            }
            catch (Exception)
            {
                throw;
            }
        }
```

File: src/vm/excep.cpp [hunk 2]
```diff
@@ -2428,9 +2428,7 @@ void StackTraceInfo::SaveStackTrace(BOOL bAllowAllocMem, OBJECTHANDLE hThrowable
...
-                    else
+                    if (!bSkipLastElement)
```
Comment (danmosemsft) [inline src/vm/excep.cpp]: Despite reading the code, I cannot form a picture of what is going on here. What is the purpose of `bSkipLastElement` ?

**Comment by danmosemsft** [comment]
Our tests for ExceptionDispatchInfo (all in `ExceptionDispatchInfoTests.netcoreapp.cs`) probably ought to do some similar line number verification ([example](https://github.com/dotnet/coreclr/issues/8521#issuecomment-265809541)). Not your job, of course.

File: src/vm/excep.cpp (hunk 2 above)
Comment (ateoi) [reply src/vm/excep.cpp]: From my analysis, this flag is set during a rethrown exception handling.
 https://github.com/dotnet/coreclr/blob/591459f32b7104a1d5c37e1a648490c8e2eb15a4/src/vm/exceptionhandling.cpp#L2555-L2558

**Comment by ateoi** [comment]
Ok, I can open a bug for the doc and submit a PR to add a test.

File: src/vm/excep.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ateoi) [reply src/vm/excep.cpp]: Pushed https://github.com/dotnet/coreclr/pull/16464/commits/d938b0c75ee1643ef5c7d49d74da45c6e61b0ad5 to remove them.

**Comment by danmosemsft** [comment]
@jkotas any other feedback?

**Comment by danmosemsft** [comment]
Oop, we generally wait for the test PR to be up before merging the implementation PR.

**Comment by jkotas** [comment]
> @jkotas any other feedback?

Looks fine to me.

> we generally wait for the test PR to be up before merging the implementation PR.

As far as I know, we have no automated tests for the stacktraces, and we have merged number of tweaks of the stacktraces without them (e.g. #14655). It would be a bit unfair to block this one on tests. I agree that we should add tests for this though.

**Comment by danmosemsft** [comment]
OK

**Pull request merged**